File: Tree/BFS/bfs.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def level_order_traversal(root: TreeNode):
    if root is None:
        return []

    res, queue = [], [root]
    while queue:
        res.append([node.val for node in queue])
        temp = []

        for node in queue:
            temp.extend([node.left, node.right])
        queue = [leaf for leaf in temp if leaf]

    return res


a = TreeNode(10)
a.left = TreeNode(20)
a.right = TreeNode(30)
a.left.left = TreeNode(40)
a.left.right = TreeNode(50)
a.right.left = TreeNode(60)
a.right.right = TreeNode(70)
print(level_order_traversal(a))

File: Tree/BFS/bfs.cpp
// level order traversal using queue
#include<bits/stdc++.h>
using namespace std;
struct node
{  int data;
   node *right, *left;
};
struct node* newnode(int key)
{  struct node* temp= new node;
   temp->data=key;
   temp->right=temp->left=NULL;
   return temp;
}
void lvl(struct node* temp)
{  if(!temp)
     return;
   queue<node*> q;
   q.push(temp);
   while(!q.empty())
   {  struct node* ptr=q.front();
      q.pop();
      cout<<ptr->data<<" ";
      if(ptr->left!=NULL) q.push(ptr->left);
      if(ptr->right!=NULL) q.push(ptr->right);
   }
} 
int main()
{  struct node* root=newnode(10);
   root->left=newnode(20);
   root->right=newnode(30);
   root->left->left=newnode(40);
   root->left->right=newnode(50);
   root->right->left=newnode(60);
   root->right->right=newnode(70);
   lvl(root);
   return 0;
}
